### ISSUE TITLE
Add result.useful and result.suppressionReasons

### ISSUE THREAD

**Issue by lgolding** [oasis-tcs/sarif-spec#268]
We propose to add two properties to the `result` object that would be useful for results management.

- Add a property `suppressionReasons` whose value is an array of strings, each of which is one of the following enumerated values with the following meanings:
    - `"falsePositive"` - The result is not actually a violation of the rule.
    - `"unshipped"` - The result occurs in code that is not shipped.
    - `"testCode"` - The result occurs in test code.
    - `"deprecated"` - The result occurs in code that is scheduled for removal.
    - `"external"` - The result occurs in code that does not belong to the development team. 

`"falsePositive"` is useful to tool authors attempting to tune their tool to maximize the signal-to-noise ratio. The other values are useful to build engineers attempting to configure their scan runs to hit only "live" code.

- Add a property `useful` whose value is a Boolean. `true` means that the engineering team found the result useful. This is a positive indicator (as opposed to the negative indicators in `suppressionReasons`). It could be used by machine learning systems to guess which results are most likely to interest the development team; it might use the learning to bring those results to the top of the list.

**Comment by michaelcfanning** [comment]
Let's start a list mail thread on this one..

**Comment by lgolding** [comment]
Adding a new enumerated value is always tricky. But unlike the never-to-be-sufficiently-discussed `threadFlowLocation.kind`, Michael and I think there is a fairly small set of values for `result.suppressionReasons` that covers the vast majority of real-world cases.

**Comment by lgolding** [comment]
I wonder whether, rather than having a single Boolean property `useful`, we should instead have a pair of integer-valued properties `usefulCount` and `notUsefulCount` to allow for up-votes and down-votes.

We could even go crazy and introduce an object with an arbitrary set of  integer-valued properties to allow for various voting options. Then the value of `result.useful` might look like this:

```
"results": [
  {
    "useful": {
      "yes": 12,
      "no": 2,
      "no opinion": 5
    }
  }
]
```

Just blue-skying here.

**Comment by sdrees** [comment]
I think the blue skies look bright and looking back, those building binaries would love such a voting hook (as long as there is a binary suppression switch as result of a voting or per convention like at the "root" of `suppressionReasons`, right? If the latter is not the case, I wonder if vendors would dare to care offering handles to those "soft" options.

**Comment by fishoak** [comment]
To the list of suppressionReasons I would add "unimportant". There are plenty of results that are simply below threshold for doing something about. Whether they are TPs or FPs may be entirely immaterial.

CodeSonar has three relevant properties that a user can set: Priority, State and Finding. Priority is just P1-P3, but also "Suppressed". State is typically one of ("assigned", "fixed", "later", "invalid", "new"), and Finding is typically one of ("FP", "TP", "Don't care", "Don't know", and "Unconfirmed"). Users can extend these lists with their own strings. These strings are entirely uninterpreted, except that the standard filters refer to them. The default filter for example, excludes FPs and Don't Care among other things.

The set of things that are filtered out by default suggest a couple of additional strings:
"invalid" might be used when someone did an analysis with the wrong configuration; "fixed" would be suitable if the defect has been corrected, but for some reason the change hasn't made it through to an analysis. (I concede that it may be bad practice to suppress a report of a "fixed" defect, but some users operate that way).

I'm intrigued by the "useful" property, but I'm not sure how I would specify it on export from CodeSonar because we don't have anything like it.

**Comment by lgolding** [comment]
`suppressionReason: unimportant` is an interesting idea. It would play nicely with the `useful` property, especially if we took my suggestion of allowing up- and down-votes. Then some dev lead might note that a result had 7 down-votes and 2 up-votes, and might then decide to suppress the result as "unimportant".

To be clear to all, this would be different from false positive. "false positive" means "the tool made a mistake". "unimportant" means "the tool got it right but I don't care".

**Comment by fishoak** [comment]
I used to give a talk about static analysis whose title was "Truth is Subjective", in which I argued that judgments of true and false positives are very strongly influenced by the role of the reviewer. While it is certainly true that tools do get it wrong sometimes, a judgment of FP on a warning is almost never solely about that, and most so for checkers that reason about semantic properties.

For example, one of my favorite demos for CodeSonar is a buffer overrun that is unmistakably a serious bug, and no sane developer would conclude otherwise. However, it's also arguably a false positive because in the particular version of the program that I analyze, the path to that bug from main() cannot be exercised. A security analyst would reasonably take that point of view because the defect is not exploitable.

Consequently it's important to think of these suppression reasons as human judgments, and not whether the tool made a mistake or not.

Wrt 'unimportant' this can also be used in situations where it is simply not worth the effort to figure out if the tool got it right or not. Again, the 'correctness' of the analysis may be entirely immaterial.
